Move oidc documentation to guides

graziang opened the pull request:
Moved OIDC documentation to guides as is.

Consider removing the paragraphs on the Node.js adapter #31326 and the Keycloak.js adapter #31327.

Closes #31329

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

graziang committed:
Move oidc documentation to guides

Closes #31329

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

mposolda commented:
@rmartinc Can you review please when you have a chance? Thanks

rmartinc commented:
@mposolda I was chatting with @graziang if separating this chapter in multiple guides makes sense (as it was done for saml with galleon and mod_auth_mellon). That way here we would need to remove the three adapter guides: JavaScript adapter, Node.js adapter and mod_auth_openidc (the three adapters would go in separate guides) and I would also remove the Java Adapters chapter completely. WDYT?

mposolda commented:
@rmartinc Yes, splitting makes sense as you suggested. Also removing Java OIDC adapters entirely makes sense. It was already communicated that OIDC java adapters were removed, so no need to have any guide with repeating this info.

rmartinc commented:
OK, then @graziang please move out those four chapters from this PR (Keycloak Java adapters, Keycloak JavaScript adapter, Keycloak Node.js adapter and mod_auth_openidc Apache HTTPD Module). And then create three new issues to have a new guide for the adapters that remain. Those three new guides would go after this guide and before saml ones (probably we need to revisit the order of guides at the end :smile: ).

graziang commented:
@rmartinc @mposolda Thanks! 

- Moved the OIDC section in the guides without the adapters.
- Removed the part about the Java adapters.
- Left the adapters in the securing apps section for now.
-  Changed all the links to OIDC guide in the documentation with a generic link to guides

Below are the three new tasks to add:

- https://github.com/keycloak/keycloak/issues/31695
- https://github.com/keycloak/keycloak/issues/31696
- https://github.com/keycloak/keycloak/issues/31697

rmartinc commented on docs/documentation/securing_apps/topics.adoc:
```diff
@@ -8,19 +8,8 @@ include::topics/overview/terminology.adoc[]
-
-include::topics/oidc/supported-grant-types.adoc[]
-
 include::topics/oidc/oidc-errors.adoc[]
```
This chapter `oid-errors` should also be included, right?

graziang replied:
done! should we do the same with saml-errors.adoc?

rmartinc replied:
Yep, we should move the saml error chapter to somewhere in the end. Probably to saml galleon chapter. But I think in this one it was clear that it should be maintained here.

rmartinc commented on docs/guides/securing-apps/partials/oidc/available-endpoints.adoc:
The title should have one less `=` sign. It's a bit strange to start with `=== Available Endpoints` third level. Just delete one `=` in each title to be aligned. In that example:

```diff
diff --git a/docs/guides/securing-apps/partials/oidc/available-endpoints.adoc b/docs/guides/securing-apps/partials/oidc/available-endpoints.adoc
index 51a287c897..b7301856c0 100644
--- a/docs/guides/securing-apps/partials/oidc/available-endpoints.adoc
+++ b/docs/guides/securing-apps/partials/oidc/available-endpoints.adoc
@@ -1,4 +1,4 @@
-=== Available Endpoints
+== Available Endpoints
 
 As a fully-compliant OpenID Connect Provider implementation, {project_name} exposes a set of endpoints that applications
 and services can use to authenticate and authorize their users.
```

graziang replied:
done! removed one level from all the titles and subtitles!

rmartinc commented on docs/guides/securing-apps/partials/oidc/fapi-support.adoc:
Same here, just delete one `=` in each title.

rmartinc commented on docs/guides/securing-apps/partials/oidc/oauth21-support.adoc:
Same here, just delete one `=` in each title.

rmartinc commented on docs/guides/securing-apps/partials/oidc/recommendations.adoc:
Same here, just delete one `=` in each title.

rmartinc commented on docs/guides/securing-apps/partials/oidc/supported-grant-types.adoc:
Same here, just delete one `=` in each title.

rmartinc reviewed (changes requested):
I think the `oidc-errors.adoc` chapter should also be moved here and in the moved chapters the titles should be aligned to second level (if not it's strange). The rest is OK, all the links are OK except one that should point to one of my chapters so I'm taking note to not forget it.

Thanks @graziang!

rmartinc reviewed (approved):
Thanks @graziang!

graziang commented:
@rmartinc I noticed that the second commit has an wrong name (i was working on the token exchange task). I think @mposolda can fix it during the merge, alternatively, I can squash the commits.

rmartinc commented:
Just squash them, the test are fast with documentation.

mposolda commented on docs/documentation/server_admin/topics/authentication/authentication-sessions.adoc:
```diff
@@ -13,13 +13,13 @@ The authentication session usually expires after 30 minutes by default. The exac
 In some rare cases, it can happen that after authentication in the first browser tab, other browser tabs are not able to restart authentication because the authentication session is already
 expired. In this case, the particular browser tab will redirect the error about the expired authentication session back to the client in a protocol specific way. For more details, see the corresponding sections
-of  link:{adapterguide_link}#_oidc-errors[OIDC documentation] and  link:{adapterguide_link}#_saml-errors[SAML documentation]. When the client application receives such an error, it can immediately resubmit the OIDC/SAML authentication request to {project_name} as
+of  *OIDC documentation* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section and  link:{adapterguide_link}#_saml-errors[SAML documentation]. When the client application receives such an error, it can immediately resubmit the OIDC/SAML authentication request to {project_name} as
```
Can we create some macro/attribute for the `https://www.keycloak.org/guides#securing-apps` ? As for the product documentation, this link will be different place.

I can see there are already some like for example https://github.com/keycloak/keycloak/blob/main/docs/documentation/topics/templates/document-attributes.adoc#L75 or https://github.com/keycloak/keycloak/blob/main/docs/documentation/topics/templates/document-attributes.adoc#L78 , so hopefully something like can be used for "Securing apps guides" as well?

I think this may apply to SAML as well as the link to "SAML documentation" on the same line is probably invalid (as AFAIK @rmartinc already moved that one to "Guides")

graziang replied:
@mposolda thanks for the review! added an attribute for  https://www.keycloak.org/guides#securing-apps in a separate commit. The SAML errors page at this link `link:{adapterguide_link}#_saml-errors` was not moved to guides @rmartinc

mposolda commented on docs/documentation/authorization_services/topics/getting-started-overview.adoc:
```diff
@@ -7,4 +7,4 @@ get started with {project_name} Authorization Services:
 * link:https://quarkus.io/guides/security-keycloak-authorization[Securing Quarkus Applications]
-* {adapterguide_link_nodejs_adapter}[Securing Node.js Applications]
\ No newline at end of file
+* *Keycloak Node.js adapter* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/authentication/authentication-sessions.adoc:
```diff
@@ -13,13 +13,13 @@ The authentication session usually expires after 30 minutes by default. The exac
+of  *OIDC documentation* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section and  link:{adapterguide_link}#_saml-errors[SAML documentation]. When the client application receives such an error, it can immediately resubmit the OIDC/SAML authentication request to {project_name} as
 this should usually automatically authenticate the user due to the existing SSO session as described earlier. As a result, the end user is authenticated automatically in all browser tabs.
-The link:{adapterguide_link}#_javascript_adapter[{project_name} Javascript adapter], link:{adapterguide_link}#_saml[{project_name} SAML adapter], and <<_identity_broker, {project_name} Identity provider>>
+The *Keycloak JavaScript adapter* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section, link:{adapterguide_link}#_saml[{project_name} SAML adapter], and <<_identity_broker, {project_name} Identity provider>>
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/authentication/flows.adoc:
```diff
@@ -343,7 +343,7 @@ claims= {
 
 The {project_name} javascript adapter has support for easy construct of this JSON and sending it in the login request.
-See link:{adapterguide_link_js_adapter}[Javascript adapter documentation] for more details.
+See *Keycloak JavaScript adapter* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section for more details.
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/authentication/x509.adoc:
```diff
@@ -191,7 +191,7 @@ image:images/x509-directgrant-flow-bindings.png[X509 Direct Grant Flow Bindings]
 
 The following example shows how to obtain an access token for a user in the realm `test` with the direct grant flow. The example is using
-link:{adapterguide_link}#_resource_owner_password_credentials_flow[OAuth2 Resource Owner Password Credentials Grant] and the confidential client `resource-owner`:
+*OAuth2 Resource Owner Password Credentials Grant* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section and the confidential client `resource-owner`:
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/clients/client-policies.adoc:
```diff
@@ -29,15 +29,15 @@ Validation of client configurations::
 
 Conformance to a required security standards and profiles such as FAPI and OAuth 2.1::
-    The _Global client profiles_ are client profiles pre-configured in {project_name} by default. They are pre-configured to be compliant with standard security profiles like link:{adapterguide_link}#_fapi-support[FAPI] and link:{adapterguide_link}#_oauth21-support[OAuth 2.1],
+    The _Global client profiles_ are client profiles pre-configured in {project_name} by default. They are pre-configured to be compliant with standard security profiles like *FAPI* and *OAuth 2.1*  in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section,
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/clients/client-policies.adoc:
```diff
@@ -29,15 +29,15 @@ Validation of client configurations::
 == Protocol
 
-The client policy concept is independent of any specific protocol. {project_name} currently supports especially client profiles for the link:{adapterguide_link}#_oidc[OpenID Connect (OIDC) protocol], but there is
+The client policy concept is independent of any specific protocol. {project_name} currently supports especially client profiles for the link:https://www.keycloak.org/guides#securing-apps[OpenID Connect (OIDC) protocol], but there is
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/clients/client-policies.adoc:
```diff
@@ -144,7 +144,7 @@ One of several purposes for this executor is to realize the security requirement
 A profile consists of several executors, which can realize a security profile like FAPI and OAuth 2.1. Profile can be configured by the Admin REST API (Admin Console) together with its executors.
 Three _global profiles_ exist and they are configured in {project_name} by default with pre-configured executors compliant with the FAPI 1 Baseline, FAPI 1 Advanced, FAPI CIBA, FAPI 2 and OAuth 2.1 specifications.
-More details exist in the FAPI and OAuth 2.1 section of the link:{adapterguide_link}#_fapi-support[{adapterguide_name}].
+More details exist in the *FAPI* and *OAuth 2.1*  in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section.
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/clients/con-protocol-mappers.adoc:
```diff
@@ -91,4 +91,4 @@ Using a lightweight access token in {project_name}::
 
 Introspection endpoint::
-  In some cases, it might be useful to trigger the token introspection endpoint with the HTTP header `Accept: application/jwt` instead of `Accept: application/json`, which can be useful especially for lightweight access tokens. See the details in the link:{adapterguide_link}#_token_introspection_endpoint[Token Introspection endpoint] section.
+  In some cases, it might be useful to trigger the token introspection endpoint with the HTTP header `Accept: application/jwt` instead of `Accept: application/json`, which can be useful especially for lightweight access tokens. See the details of *Token Introspection endpoint* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section.
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/clients/oidc/con-audience.adoc:
```diff
@@ -103,16 +103,16 @@ image:images/audience_mapper.png[]
 See:
 +
-** link:{adapterguide_link}#_javascript_adapter[javascript adapter section] if your application uses the javascript adapter.
+** *Keycloak JavaScript adapter* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section if your application uses the javascript adapter.
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/clients/proc-evaluating-client-scopes.adoc:
```diff
@@ -15,7 +15,7 @@ image:images/client-scopes-evaluate.png[]
 [NOTE]
 ====
-To send a custom value for a *scope* parameter from your application, see the link:{adapterguide_link_latest}#_javascript_adapter[javascript adapter section], for javascript adapters.
+To send a custom value for a *scope* parameter from your application, see the *Keycloak JavaScript adapter* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section, for javascript adapters.
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/sso-protocols/con-oidc-auth-flows.adoc:
```diff
@@ -89,7 +89,7 @@ It is possible to specify that the refresh token is considered invalid once it i
 
 {project_name} also supports the situation that no refresh token rotation exists. In this case, a refresh token is returned during login, but subsequent responses from refresh-token requests will not
-return new refresh tokens. This practice is recommended for instance in the link:{adapterguide_link}#_fapi-support[FAPI 2 draft specification].
+return new refresh tokens. This practice is recommended for instance in the *FAPI 2 draft specification* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section.
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/sso-protocols/con-oidc-auth-flows.adoc:
```diff
@@ -107,11 +107,11 @@ This is used by clients running on internet-connected devices that have limited
 An administrator can configure Client Initiated Backchannel Authentication (CIBA) related operations as `CIBA Policy` per realm.
 
-Also please refer to other places of {project_name} documentation like link:{adapterguide_link}#_backchannel_authentication_endpoint[Backchannel Authentication Endpoint section] of {adapterguide_name} and link:{adapterguide_link}#_client_initiated_backchannel_authentication_grant[Client Initiated Backchannel Authentication Grant section] of {adapterguide_name}.
+Also please refer to other places of {project_name} documentation like *Backchannel Authentication Endpoint* and *Client Initiated Backchannel Authentication Grant* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section.
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/sso-protocols/oidc.adoc:
```diff
@@ -109,11 +109,11 @@ This is used by clients running on internet-connected devices that have limited
 An administrator can configure Client Initiated Backchannel Authentication (CIBA) related operations as `CIBA Policy` per realm.
 
-Also please refer to other places of {project_name} documentation like link:{adapterguide_link}#_backchannel_authentication_endpoint[Backchannel Authentication Endpoint section] of {adapterguide_name} and link:{adapterguide_link}#_client_initiated_backchannel_authentication_grant[Client Initiated Backchannel Authentication Grant section] of {adapterguide_name}.
+Also please refer to other places of {project_name} documentation like *Backchannel Authentication Endpoint* and *Client Initiated Backchannel Authentication Grant* in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section.
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/threat/fapi-compliance.adoc:
```diff
@@ -2,6 +2,6 @@
 
 To make sure that {project_name} server will validate your client to be more secure and FAPI compliant, you can configure client policies
-for the FAPI support. Details are described in the FAPI section of link:{adapterguide_link}#_fapi-support[{adapterguide_name}]. Among other things, this ensures some security best practices
+for the FAPI support. *FAPI* details are described in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section. Among other things, this ensures some security best practices
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda commented on docs/documentation/server_admin/topics/threat/oauth21-compliance.adoc:
```diff
@@ -2,4 +2,4 @@
 
 To make sure that {project_name} server will validate your client to be more secure and OAuth 2.1 compliant, you can configure client policies
-for the OAuth 2.1 support. Details are described in the OAuth 2.1 section of link:{adapterguide_link}#_oauth21-support[{adapterguide_name}].
+for the OAuth 2.1 support. *OAuth 2.1* details are described in the link:https://www.keycloak.org/guides#securing-apps[securing apps] section.
```
Same like my other comment (can be hopefully attribute instead of hardcoded link).

mposolda reviewed (commented):
@graziang @rmartinc Thanks for the docs and review!

I have 16 comments, but all are for the same thing: was just wondering if we can have "attribute" instead of having hardcoded link to Same like my other comment (can be hopefully attribute instead of hardcoded link to https://www.keycloak.org/guides#securing-apps everywhere as this will differ in the product documentation).

graziang committed:
external link for securing apps

Closes #31329

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

mposolda reviewed (approved):
@graziang Thanks for the update!
